### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by brifino**
Expanded and added functionality----- 
    Display Class
    Library Class

Added Track Class functions ----- These are just attempts. But hopefully they can point us in the right direction

     setWriteFileName - Sets rearFileName to one selected from Library screen. 
     
     reverse - reverses a file. 

Changed a few things in main, but just for Display and Library class interactions and calls

### CODE REVIEW

File: app/lib/library/include/library.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#ifndef LIBRARY_HPP
+#define LIBRARY_HPP
+
+#include <Arduino.h> // Font for display
+
+class Library{
+
+    private:
+        String * fileArray;
+        static const int SIZE = 20;
```
Comment (dieboljo) [inline app/lib/library/include/library.hpp:10]: We should look into dynamically creating this array. If there are more than 20 files in the directory, this will cause the application to crash. We could look into getting the file count before initializing the array?

File: app/lib/library/src/library.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#include <library.hpp>
+#include <Arduino.h>
+#include <SD.h>
+
+
+const int SDCARD_MOSI_PIN = 11;
+const int SDCARD_SCK_PIN = 13;
+const int SDCARD_CS_PIN = BUILTIN_SDCARD;
```
Comment (dieboljo) [inline app/lib/library/src/library.cpp:8]: These aren't used in this file.

File: app/lib/library/src/library.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#include <SD.h>
+
+
+const int SDCARD_MOSI_PIN = 11;
+const int SDCARD_SCK_PIN = 13;
+const int SDCARD_CS_PIN = BUILTIN_SDCARD;
+
+// create the array of filenames
+void Library::array(){
+    File root = SD.open("/");
```
Comment (dieboljo) [inline app/lib/library/src/library.cpp:12]: Reading from the root directory displays the files that are being used as "file buffers", which shouldn't be accessible. We should probably use a different subdirectory, instead. In the branch `feature/save-loop`, a _loops_ subdirectory is used. In any case, we would just check if the directory exists, and create it if not.

File: app/lib/library/src/library.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include <Arduino.h>
+#include <SD.h>
+
+
+const int SDCARD_MOSI_PIN = 11;
+const int SDCARD_SCK_PIN = 13;
+const int SDCARD_CS_PIN = BUILTIN_SDCARD;
+
+// create the array of filenames
+void Library::array(){
```
Comment (dieboljo) [inline app/lib/library/src/library.cpp:11]: How about naming this `setup` or `begin` for clarity?

File: app/lib/track/src/track.cpp [hunk 5]
```diff
@@ -7,6 +7,9 @@
 const size_t Track::playBufferSize = 65536;    // 64k
 const size_t Track::recordBufferSize = 131072; // 128k
 
+// For reverse track ---- NEEDED?
+#define BUFFER_SIZE 1024
```
Comment (dieboljo) [inline app/lib/track/src/track.cpp:11]: Like I mentioned earlier, this doesn't need to be a `#define` macro. Also, the name `BUFFER_SIZE` is a bit vague, especially since this class has `playBufferSize` and `recordBufferSize` variables.

File: app/lib/track/src/track.cpp [hunk 6]
```diff
@@ -269,3 +272,58 @@ Track::Track(const char *f1, const char *f2, AudioInputI2S *s)
...
+  Serial.print(Length);
+  char newReadFile[Length];
+  
+  name.toCharArray(newReadFile, Length);
+  Serial.printf("Read file name was: %s\n", readFileName);
+  this->readFileName = newReadFile;
+  Serial.printf("Read file name now: %s\n", readFileName);
+  Serial.printf("Write file name: %s\n", writeFileName);
+ 
+}
```
Comment (dieboljo) [inline app/lib/track/src/track.cpp:289]: The `readFileName` and `writeFileName` values should stay constant, and when a loop is loaded, it should be copied over to the file labelled by `readFileName`, not `writeFileName`. Currently, the program would crash if the same loop was loaded into two separate tracks. Also, copying the file allows the original to remain as a backup.

File: app/lib/display/include/display.hpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        void setRevBool(boolean change) {this->reverseBool = change;}
+        boolean getRevBool() const {return this->reverseBool;}
+
+        void Setup();
+
+        void Boot();
+
+        void mainScreen();
+
+        void Library();
```
Comment (dieboljo) [inline app/lib/display/include/display.hpp:170]: This method is not implemented.

File: app/lib/display/include/display.hpp [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+        void setNameChange(boolean change) {this->nameChange = change;}
+        boolean getNameChange() const {return this->nameChange;}
+
+        void setRevBool(boolean change) {this->reverseBool = change;}
+        boolean getRevBool() const {return this->reverseBool;}
+
+        void Setup();
+
+        void Boot();
```
Comment (dieboljo) [inline app/lib/display/include/display.hpp:166]: Member functions should start with a lowercase letter.

File: app/src/main.cpp [hunk 9]
```diff
@@ -65,6 +71,18 @@ void setup() {
...
+  // create fileArray containing names of SD card files
+  String * fileArray = lib.returnValue();
+  //copy fileArray to display array
+  disp.setArray(fileArray);
+
+  // boot display up
+  disp.Setup();
+  disp.Boot();
+  delay(4000);
+  disp.mainScreen();
```
Comment (dieboljo) [inline app/src/main.cpp:84]: Can't these all be moved into the class's setup function?

File: app/src/main.cpp [hunk 10]
```diff
@@ -83,6 +101,44 @@ void loop() {
 
   monitorAudioEngine();
 
+  // Run display controls
+  disp.displayVol();
```
Comment (dieboljo) [inline app/src/main.cpp:105]: The volume is already checked and adjusted in the `adjustVolume` button. Could we move this call into that function, and just display the new volume instead of checking it again?

File: app/src/main.cpp [hunk 11]
```diff
@@ -83,6 +101,44 @@ void loop() {
...
+    }
+  }
+  
+  // Moinitor reverse flag
+  if(disp.getRevBool()){
+    Serial.println("Reverse Entered");
+    //track.reverse();        Commented out - Need work
+    disp.setRevBool(false);
+    status = Status::Play;
+  }
```
Comment (dieboljo) [inline app/src/main.cpp:137]: Until reverse is working, let's remove the related code from this pull request. It can be included in a separate PR.

File: app/src/main.cpp [hunk 12]
```diff
@@ -93,25 +149,42 @@ void loop() {
...
+  */
 
   if (buttons.record.fallingEdge()) {
     Serial.println("Record Button Pressed");
     switch (status) {
     case Status::Record:
+      disp.SetRecordButton(false);
+      disp.SetStopButton(false);
+      disp.SetPlayButton(true);
+      disp.displayTrack("Playing");
```
Comment (dieboljo) [inline app/src/main.cpp:161]: Mentioned earlier, this can all be accomplished with a single function call.

File: app/src/main.cpp [hunk 13]
```diff
@@ -83,6 +101,44 @@ void loop() {
 
   monitorAudioEngine();
 
+  // Run display controls
+  disp.displayVol();
+  disp.displayPan();
```
Comment (dieboljo) [inline app/src/main.cpp:106]: Similar to the above, the "pan" value is already updated to a value between 0 and 1 in the `adjustPan` function above. So this function shouldn't need to read the knob again, and it can just read the global `pan` variable. This way, if we change the logic for reading the knob, it will only need to be changed in one place.

File: app/src/main.cpp [hunk 14]
```diff
@@ -83,6 +101,44 @@ void loop() {
...
+      mode = Mode::Replace;
+      disp.setModeChange(0);
+      Serial.println("Mode: REPLACE");
+    }
+    else{
+      mode = Mode::Overdub;
+      disp.setModeChange(0);
+      Serial.println("Mode: OVERDUB");
+    }
+  }
```
Comment (dieboljo) [inline app/src/main.cpp:129]: For consistency across the application, why not use the `Mode` enum in these functions instead of these arbitrary integers?

**Review by dieboljo — CHANGES_REQUESTED** [review]
The display is coming together, and the touch screen works better than it did in the example that came with the board! I do have a number of suggestions, though, many of which I've detailed below in code comments.

Other general suggestions:
- In the design document, we had discussed the touchscreen and opted not to use it. If we are going to use it, the UI should make it clear what is a control and what is a displayed value. Currently, everything is styled as a button, when the only tap-able items are mode, reverse, and "show library". I would suggest removing the borders and backgrounds from all other items.
- The `displayTrack` function doesn't seem to display the track at all, and the string it displays seems a bit arbitrary. I think the part of the UI this controls should always be a "Go to library" button, and shouldn't have anything to do with the current status. More on that in the code comments.
- Something is blocking the main loop, and button presses are not being read as they should.
- It would help readability if the verbs used in function names were consistent. Usually, "set" is used for a function that updates a private variable, but in the `Display` class, both "set" and "display" are used in functions that draw a UI component. Whatever is used, consistency is most important. To me, "draw" makes most sense for a function that draws a component, and "show" or "display" for switching to a new screen.

File: app/lib/display/include/display.hpp [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+
+// Define Reverse button location and size
+#define REVERSE_X 160 
+#define REVERSE_Y 160
+#define REVERSE_W 60
+#define REVERSE_H 60
+
+
+//Font for buttons
+#define BUTTON_FONT Arial_14
```
[the remark this answers is not in the document]
Comment (brifino) [reply app/lib/display/include/display.hpp:135]: Removed all the #define statements. Now using a struct as recommended

File: app/lib/display/src/display.cpp [hunk 16]
```diff
@@ -0,0 +1,487 @@
+#include <display.hpp>
+#include <Arduino.h>
+#include <SD.h>
+
+ILI9341_t3 tft = ILI9341_t3(TFT_CS, TFT_DC);
+XPT2046_Touchscreen ts(TS_CS);
+TS_Point p;
```
[the remark this answers is not in the document]
Comment (brifino) [reply app/lib/display/src/display.cpp:7]: moved into Display class

File: app/src/main.cpp (hunk 12 above)
Comment (brifino) [reply app/src/main.cpp]: fixed

File: app/src/main.cpp (hunk 11 above)
Comment (brifino) [reply app/src/main.cpp:132]: sounds good!
